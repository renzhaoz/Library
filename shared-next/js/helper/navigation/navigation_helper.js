const DIR = Object.freeze({
  LEFT: 0,
  RIGHT: 1,
  UP: 2,
  DOWN: 3
});

(function navigationHelper(exports) {
  const NavigationHelper = {
    controls: null,

    resetMenu() {
      const selector = 'form[data-subtype=menu]';
      const postfix = '>button.menu-button';
      const menuform = document.querySelector(selector);
      for (let i = 1; i < menuform.children.length; i++) {
        const subid = menuform.children[i].id;
        if (subid && subid.indexOf('submenu_') === 0) {
          this.reset(
            // eslint-disable-next-line no-use-before-define
            VerticalNavigator,
            () => {
              return document.querySelectorAll(`#${subid}${postfix}`);
            },
            0,
            subid
          );
        }
      }
      return this.reset(
        // eslint-disable-next-line no-use-before-define
        VerticalNavigator,
        () => {
          return document.querySelectorAll(`#mainmenu${postfix}`);
        },
        0,
        'mainmenu'
      );
    },

    // eslint-disable-next-line max-params
    reset(navigator, getControls, getFocusIndex, navPrefix, notSetFocus) {
      if (!navigator) {
        return null;
      }
      if (typeof getControls !== 'function') {
        return null;
      }
      this.controls = getControls();
      if (!this.controls || this.controls.length === 0) {
        return null;
      }
      navigator.count = this.controls.length;
      this.updateNav(navigator, navPrefix);

      const focusIndex =
        typeof getFocusIndex === 'function' ? getFocusIndex() : 0;
      const focusedElement = this.controls[focusIndex];
      if (!notSetFocus) this.setFocus(focusedElement);
      return focusedElement;
    },

    setFocus(element) {
      if (!element) return;
      const focused = document.querySelectorAll('.focus');
      if (focused.length > 0) {
        focused[0].classList.remove('focus');
      }
      element.setAttribute('tabindex', 1);
      element.classList.add('focus');
      if ('undefined' !== typeof inputHandler)
        // eslint-disable-next-line no-undef
        inputHandler.focusChanged(element);
      else element.focus();
    },

    updateNav(navigator, navPrefix) {
      for (let i = 0; i < this.controls.length; i++) {
        const control = this.controls[i];
        control.setAttribute('data-nav-id', str(i));
        control.style.setProperty('--nav-left', str(navigator.left(i)));
        control.style.setProperty('--nav-right', str(navigator.right(i)));
        control.style.setProperty('--nav-up', str(navigator.up(i)));
        control.style.setProperty('--nav-down', str(navigator.down(i)));
        control.setAttribute('tabindex', 0);
      }
      function str(navIndex) {
        if (navPrefix === undefined) navPrefix = '';
        return navPrefix + navIndex;
      }
    },

    elementIsVisible(element, top, bottom) {
      let visible = true;
      if (element.offsetWidth === 0 || element.offsetHeight === 0)
        return visible;
      const rects = element.getClientRects();
      for (const rect of rects) {
        if (rect.bottom + bottom > window.innerHeight || rect.top < top) {
          visible = false;
          break;
        }
      }
      return visible;
    },

    // Gaia-header: 50px, softkey: 40px, (ignore statusbar: 30px)
    // eslint-disable-next-line max-params
    scrollToElement(element, event, top, bottom) {
      if (!this.elementIsVisible(element, top, bottom)) {
        if (event) {
          if (event.key === 'ArrowDown' || event.key === 'ArrowUp')
            element.scrollIntoView(false);
        } else {
          element.scrollIntoView(false);
        }
      }
    },

    Key2Dir(key) {
      // eslint-disable-next-line init-declarations
      let dir;
      switch (key) {
        case 'ArrowLeft':
          dir = DIR.LEFT;
          break;
        case 'ArrowRight':
          dir = DIR.RIGHT;
          break;
        case 'ArrowUp':
          dir = DIR.UP;
          break;
        case 'ArrowDown':
          dir = DIR.DOWN;
          break;
        default:
          break;
      }
      return dir;
    }
  };

  // Horizontal Navigator
  const HorizontalNavigator = {
    count: 0,
    left(navIndex) {
      return 0 === navIndex ? this.count - 1 : navIndex - 1;
    },
    right(navIndex) {
      return this.count - 1 === navIndex ? 0 : navIndex + 1;
    },
    up(navIndex) {
      return navIndex;
    },
    down(navIndex) {
      return navIndex;
    }
  };

  // Vertical Navigator
  const VerticalNavigator = {
    count: 0,
    left(navIndex) {
      return navIndex;
    },
    right(navIndex) {
      return navIndex;
    },
    up(navIndex) {
      return 0 === navIndex ? this.count - 1 : navIndex - 1;
    },
    down(navIndex) {
      return this.count - 1 === navIndex ? 0 : navIndex + 1;
    }
  };

  // Box Navigator
  const BoxNavigator = {
    count: 0,
    columnCount: 2,
    left(navIndex) {
      return navIndex > 0 && navIndex < this.count
        ? --navIndex
        : this.count - 1;
    },
    right(navIndex) {
      return navIndex >= 0 && navIndex < this.count - 1 ? ++navIndex : 0;
    },
    up(navIndex) {
      navIndex -= this.columnCount;
      if (navIndex < 0) {
        navIndex += this.columnCount * Math.ceil(this.count / this.columnCount);
        const remainder = this.count % this.columnCount;
        if (
          (navIndex === this.count ||
            (navIndex === this.count + 1 && remainder === 1)) &&
          remainder !== 0
        )
          navIndex = this.count - 1;
        if (navIndex > this.count) navIndex -= this.columnCount;
      }
      return navIndex;
    },
    down(navIndex) {
      navIndex += this.columnCount;
      const remainder = this.count % this.columnCount;
      if (
        (navIndex === this.count ||
          (navIndex === this.count + 1 && remainder === 1)) &&
        remainder !== 0
      )
        navIndex = this.count - 1;
      if (navIndex > this.count || (navIndex === this.count && remainder === 0))
        navIndex %= this.columnCount;
      return navIndex;
    }
  };

  // Group Navigator
  const GroupNavigator = {
    count: 0,
    columnCount: 3,
    getGroupInfo: null,
    left(navIndex) {
      return navIndex > 0 && navIndex < this.count
        ? --navIndex
        : this.count - 1;
    },
    right(navIndex) {
      return navIndex >= 0 && navIndex < this.count - 1 ? ++navIndex : 0;
    },
    up(navIndex) {
      let index = -1;
      if (typeof this.getGroupInfo !== 'function') {
        return index;
      }
      const groupInfo = this.getGroupInfo(navIndex, true);
      const curIdx = groupInfo.curGroupIndex;
      const destLen = groupInfo.destGroupLen;
      if (curIdx >= this.columnCount) {
        // In the current group
        index = navIndex - this.columnCount;
      } else {
        // In the destination group
        index = navIndex - curIdx;
        if (0 === index) index = this.count;
        let lastRowCount = destLen % this.columnCount;
        if (0 === lastRowCount) lastRowCount = this.columnCount;
        index -= lastRowCount;
        const pos = curIdx % this.columnCount;
        if (pos < lastRowCount) {
          index += pos;
        } else if (destLen > this.columnCount) {
          // More than 1 row
          index = index - this.columnCount + pos;
        } else {
          index += lastRowCount - 1;
        }
      }
      return index;
    },

    down(navIndex) {
      let index = -1;
      const groupInfo = this.getGroupInfo(navIndex, false);
      const curIdx = groupInfo.curGroupIndex;
      const curLen = groupInfo.curGroupLen;
      if (curIdx + this.columnCount < curLen) {
        // In the current group
        index = navIndex + this.columnCount;
      } else {
        // In the destination group
        index = navIndex + (curLen - curIdx);
        if (this.count === index) index = 0;
        const pos = curIdx % this.columnCount;
        index += Math.min(pos, groupInfo.destGroupLen - 1);
      }
      return index;
    }
  };

  exports.NavigationHelper = NavigationHelper;
  exports.HorizontalNavigator = HorizontalNavigator;
  exports.VerticalNavigator = VerticalNavigator;
  exports.BoxNavigator = BoxNavigator;
  exports.GroupNavigator = GroupNavigator;
})(window);
