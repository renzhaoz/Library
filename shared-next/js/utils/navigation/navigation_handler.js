/* global NavigationMap LazyLoader */
(function navigationHandler() {
  const loader = LazyLoader;
  // App should hold this file in /js folder
  loader.load('js/navigation_map.js', () => {
    NavigationMap.init();
  });
  window.addEventListener('keydown', evt => {
    handleKeydown(evt);
  });

  window.addEventListener('menuEvent', event => {
    if (NavigationMap) {
      NavigationMap.optionMenuVisible = event.detail.menuVisible;
    }
  });

  function handleKeydown(evt) {
    let el = evt.target;
    let bestElementToFocus = null;
    if (
      NavigationMap &&
      (NavigationMap.currentActivatedLength > 0 || NavigationMap.lockNavigation)
    ) {
      return;
    }
    if (
      document.activeElement.type === 'select-one' ||
      document.activeElement.type === 'date'
    ) {
      return;
    }
    if (evt.key === 'Enter' || evt.key === 'Accept') {
      handleClick(evt);
    } else {
      if (!evt.target.classList) {
        return;
      }
      if (!evt.target.classList.contains('focus')) {
        el = document.querySelector('.focus');
      }
      bestElementToFocus = findElementFromNavProp(el, evt);
      if (bestElementToFocus !== null) {
        if (
          evt.key === 'ArrowDown' &&
          evt.repeat &&
          bestElementToFocus.offsetTop < el.offsetTop
        ) {
          evt.preventDefault();
          return;
        }
        if (
          evt.key === 'ArrowUp' &&
          evt.repeat &&
          bestElementToFocus.offsetTop > el.offsetTop
        ) {
          evt.preventDefault();
          return;
        }
        const prevFocused = document.querySelectorAll('.focus');
        if (bestElementToFocus === prevFocused[0]) {
          return;
        }
        if (prevFocused.length > 0) {
          prevFocused[0].classList.remove('focus');
        }
        if (NavigationMap.scrollToElement === undefined) {
          bestElementToFocus.scrollIntoView(false);
        } else {
          NavigationMap.scrollToElement(bestElementToFocus, evt);
        }
        bestElementToFocus.classList.add('focus');
        if (NavigationMap.ignoreFocus === null || !NavigationMap.ignoreFocus) {
          bestElementToFocus.focus();
          /*
           * App make decision to focus on next element.
           * Then, we have to prevent default here.
           */
          evt.preventDefault();
        }
        document.dispatchEvent(
          new CustomEvent('focusChanged', {
            detail: {
              focusedElement: bestElementToFocus
            }
          })
        );
      }
    }
  }

  function findElementFromNavProp(currentlyFocused, evt) {
    let elementID = null;
    if (currentlyFocused === null) {
      return null;
    }
    if (
      NavigationMap !== undefined &&
      NavigationMap.disableNav !== undefined &&
      NavigationMap.disableNav
    )
      return null;
    const elmStyle = currentlyFocused.style;
    const isRtl = document.dir === 'rtl';
    let { key } = evt;
    const keyMap = {
      ArrowLeft: '--nav-left',
      ArrowRight: '--nav-right',
      ArrowUp: '--nav-up',
      ArrowDown: '--nav-down',
      Home: '--nav-home',
      MozHomeScreen: '--nav-home'
    };

    switch (key) {
      case 'ArrowLeft':
      case 'ArrowRight':
        if (isRtl) {
          key = key === 'ArrowLeft' ? 'ArrowRight' : 'ArrowLeft';
        }
        elementID = elmStyle.getPropertyValue(keyMap[key]);
        break;
      case 'ArrowUp':
      case 'ArrowDown':
      case 'Home':
      case 'MozHomeScreen':
        elementID = elmStyle.getPropertyValue(keyMap[key]);
        break;
      default:
        break;
    }

    if (!elementID) {
      return null;
    }
    const selector = `[data-nav-id="${elementID}"]`;
    return document.querySelector(selector);
  }

  function handleClick(evt) {
    const el = document.querySelector('.focus');
    if (el) {
      el.focus();
    }

    if (
      NavigationMap &&
      NavigationMap.optionMenuVisible &&
      !evt.target.classList.contains('menu-button')
    ) {
      // Workaround for case of quick click just right after option menu opening start
      const selectedMenuElement = document.querySelector(
        'menu button.menu-button'
      );
      if (selectedMenuElement && selectedMenuElement.click) {
        selectedMenuElement.click();
      }
    } else if (NavigationMap && NavigationMap.handleClick) {
      // Custimization of click action.
      NavigationMap.handleClick(evt);
    } else {
      evt.target.click();
      for (let i = 0; i < evt.target.children.length; i++) {
        evt.target.children[i].click();
      }
    }
  }
})();
