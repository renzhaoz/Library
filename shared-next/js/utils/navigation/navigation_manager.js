/* global NavigationMap */
(function navigationManager(exports) {
  let _focusedEl = null; // For saving current focus element
  const isFunction = function isFunction(obj) {
    return !!(obj && obj.constructor && obj.call && obj.apply);
  };

  const NavigationManager = function NavigationManager() {
    this.context = {};
    this.currentSelector = undefined;
    this.nextId = this.generatorId(0);

    /*
     * For backward compatibility
     * some applications do not invoke init method and still use navigation_handler.js
     * TODO: move to init once navigation_handler.js has removed from Dialer and CallScreen
     */
    document.addEventListener(
      'focusChanged',
      this.focusChandedHandler.bind(this)
    );
  };

  NavigationManager.prototype.CLASS_NAME = 'NavigationManager';
  NavigationManager.prototype.DEBUG = false;

  NavigationManager.prototype.init = function init(config) {
    this.config = config || {
      clickHandler: undefined
    };

    window.addEventListener('keydown', this.keyHandler.bind(this));
  };

  NavigationManager.prototype.keyHandler = function keyHandler(evt) {
    let el = evt.target;
    let focusElement = null;
    let prevFocused = null;

    this.debug(`key ${evt.key}`);
    if (
      NavigationMap &&
      NavigationMap.currentActivatedLength &&
      NavigationMap.currentActivatedLength > 0
    ) {
      return;
    }
    if (!el) {
      return;
    }

    if (evt.key === 'Enter' || evt.key === 'Accept') {
      this.click(el);
      this.dispatchEvent('accepted', {
        acceptedEl: el
      });
    } else {
      if (!el.classList) {
        return;
      }
      if (!el.classList.contains('focus')) {
        this.debug('event target does not have focus');
        el = document.querySelector('.focus');
      }

      focusElement = this.findElementFromNavProp(el, evt);
      if (focusElement) {
        prevFocused = document.querySelectorAll('.focus');
        if (focusElement === prevFocused[0]) {
          return;
        }
        if (prevFocused.length) {
          prevFocused[0].classList.remove('focus');
        }
        if (isFunction(this.scrollToElement)) {
          this.scrollToElement(focusElement, evt);
        } else {
          focusElement.scrollIntoView(false);
        }
        focusElement.classList.add('focus');
        _focusedEl = focusElement;
        const inPutForT9 = focusElement.querySelector('input[type="text"]');

        if (inPutForT9) {
          inPutForT9.focus();
        } else {
          focusElement.focus();
        }
        this.dispatchEvent('focusChanged', {
          focusedElement: focusElement
        });
      }
    }
  };

  NavigationManager.prototype.findElementFromNavProp = function findElementFromNavProp(
    currentlyFocused,
    evt
  ) {
    if (!currentlyFocused) {
      return null;
    }

    let elementID = null;
    const elmStyle = currentlyFocused.style;
    let handled = true;

    switch (evt.key) {
      case 'ArrowLeft':
        elementID = elmStyle.getPropertyValue('--nav-left');
        break;
      case 'ArrowRight':
        elementID = elmStyle.getPropertyValue('--nav-right');
        break;
      case 'ArrowUp':
        elementID = elmStyle.getPropertyValue('--nav-up');
        break;
      case 'ArrowDown':
        elementID = elmStyle.getPropertyValue('--nav-down');
        break;
      case 'Home':
      case 'MozHomeScreen':
        elementID = elmStyle.getPropertyValue('--nav-home');
        break;
      default:
        handled = false;
    }
    if (!elementID) {
      return null;
    }
    if (handled) {
      evt.preventDefault();
    }
    const selector = `[data-nav-id="${elementID}"]`;
    return document.querySelector(selector);
  };

  NavigationManager.prototype.click = function click(el) {
    if (this.config.clickHandler) {
      this.config.clickHandler(el);
    } else {
      el.click();
    }
  };

  NavigationManager.prototype.dispatchEvent = function dispatchEvent(
    event,
    data
  ) {
    const dataTemp =
      (data && {
        detail: data
      }) ||
      {};
    document.dispatchEvent(new CustomEvent(event, dataTemp));
  };

  NavigationManager.prototype.focusChandedHandler = function focusChandedHandler(
    e
  ) {
    const focusElement = e.detail.focusedElement,
      id = focusElement && focusElement.getAttribute('data-nav-id');

    if (!focusElement) return;

    if (this.currentSelector) {
      this.context[this.currentSelector] = id;
    }
  };

  NavigationManager.prototype.generatorId = function generatorId(base) {
    let localId = base;
    return function() {// eslint-disable-line
      return localId++;
    };
  };

  NavigationManager.prototype.delNavId = function delNavId(selector) {
    const items = document.querySelectorAll(selector);
    if (!items.length) {
      return;
    }

    Array.prototype.forEach.call(items, item =>
      item.removeAttribute('data-nav-id')
    );
  };

  NavigationManager.prototype.initNavId = function initNavId(items) {
    Array.prototype.forEach.call(items, item => {
      const navId = item.getAttribute('data-nav-id');
      if (!navId) {
        item.setAttribute('data-nav-id', this.nextId());
      }
    });
  };

  NavigationManager.prototype.prepareElements = function prepareElements(
    selector
  ) {
    if (!selector) {
      throw Error('selector is undefined');
    }

    const items = document.querySelectorAll(selector);
    const focused = Array.prototype.slice.call(
      document.querySelectorAll('.focus')
    );
    const focusedLength = focused.length;

    this.currentSelector = selector;
    this.debug(`items.length ${items.length}`);
    if (!items.length) {
      this.dispatchEvent('focusChanged', {
        focusChanged: null
      });
      return null;
    }
    if (focusedLength) {
      for (let j = 0; j < focusedLength; j++) {
        focused[j].classList.remove('focus');
      }
    }

    this.initNavId(items);
    return items;
  };

  // eslint-disable-next-line
  NavigationManager.prototype.reset = function reset(
    selector,
    navId,
    direction,
    withoutUpdate
  ) {
    const items = this.prepareElements(selector);
    _focusedEl = null; // Clear current focused cache
    if (items) {
      let [item] = items;
      let el = null;
      if (navId) {
        el = document.querySelector(`[data-nav-id="${navId}"]`);
        if (el) {
          item = el;
        }
      }
      this.context[selector] = item.dataset.navId;
      this.setFocus(item);
      if (!withoutUpdate) {
        this.update(selector, direction);
      }
    }
  };

  NavigationManager.prototype.unfocus = function unfocus() {
    const el = Array.prototype.slice.call(document.querySelectorAll('.focus'));
    el.forEach(item => {
      item.classList.remove('focus');
    });
  };
  NavigationManager.prototype.resetByNode = function resetByNode(
    selector,
    node,
    direction
  ) {
    const items = this.prepareElements(selector);
    _focusedEl = null;
    if (items) {
      let [item] = items;
      let arrElements = null;
      let checkNodeInSelector = null;
      if (node) {
        arrElements = Array.prototype.slice.call(
          document.querySelectorAll(selector)
        );
        checkNodeInSelector = arrElements.some(function check(elem) {
          return elem === node;
        });
        if (!checkNodeInSelector) {
          console.error('node is not exist in selector');
          return;
        }
        item = node;
      }
      this.context[selector] = item.dataset.navId;
      this.setFocus(item);
      this.update(selector, direction);
    }
  };

  /*
   * TODO: this code is specific. Further clients should implement update logic and
   * pass it to Navigation Manager
   */
  NavigationManager.prototype.update = function update(selector, direction) {
    let item = null;
    let j = 0;
    let nextItem = null;
    let prevItem = null;

    if (!document.querySelectorAll('.focus').length) {
      this.reset(selector);
      return;
    }
    if (!selector) {
      throw Error('selector is undefined');
    }

    const items = document.querySelectorAll(selector);

    this.debug(`items.length ${items.length}`);
    this.initNavId(items);

    if (items.length <= 1) {
      const index = 0;
      if ((item = items[index])) {
        item.style.removeProperty('--nav-down');
        item.style.removeProperty('--nav-up');
      }
      return;
    }
    this.debug(`${selector}.length = ${items.length}`);
    if (!direction) {
      direction = 'vertical';
    }

    if (typeof this.doUpdateCb === 'function') {
      this.doUpdateCb(items);
      return;
    }

    switch (direction) {
      case 'vertical':
        for (j = 0; j < items.length; j++) {
          item = items[j];
          prevItem = items[j - 1] || items[items.length - 1];
          nextItem = items[j + 1] || items[0];

          item.style.setProperty(
            '--nav-down',
            nextItem.getAttribute('data-nav-id')
          );
          item.style.setProperty(
            '--nav-up',
            prevItem.getAttribute('data-nav-id')
          );
        }
        break;
      case 'horizontal':
        for (j = 0; j < items.length; j++) {
          item = items[j];
          prevItem = items[j - 1] || items[items.length - 1];
          nextItem = items[j + 1] || items[0];

          item.style.setProperty(
            '--nav-right',
            nextItem.getAttribute('data-nav-id')
          );
          item.style.setProperty(
            '--nav-left',
            prevItem.getAttribute('data-nav-id')
          );
        }
        break;
      default:
        break;
    }
  };

  NavigationManager.prototype._linkToUp = function _linkToUp(from, to) {
    from.style.setProperty('--nav-up', to.getAttribute('data-nav-id'));
  };

  NavigationManager.prototype._linkToDown = function _linkToDown(from, to) {
    from.style.setProperty('--nav-down', to.getAttribute('data-nav-id'));
  };

  NavigationManager.prototype._linkToLeft = function _linkToLeft(from, to) {
    from.style.setProperty('--nav-left', to.getAttribute('data-nav-id'));
  };

  NavigationManager.prototype._linkToRight = function _linkToRight(from, to) {
    from.style.setProperty('--nav-right', to.getAttribute('data-nav-id'));
  };

  NavigationManager.prototype.appendToList = function appendToList(
    listContainer,
    items,
    direction
  ) {
    const len = listContainer.children.length;
    const itemsLen = items.length;

    items.forEach(item => {
      listContainer.appendChild(item);
      item.setAttribute('data-nav-id', this.nextId());
    });

    let linkToNxt = null;
    let linkToPrv = null;
    direction = direction || 'vertical';

    switch (direction) {
      case 'vertical':
        linkToPrv = this._linkToUp;
        linkToNxt = this._linkToDown;
        break;

      case 'horizontal':
        linkToPrv = this._linkToLeft;
        linkToNxt = this._linkToRight;
        break;
      default:
        break;
    }

    if (len) {
      linkToPrv(items[0], listContainer.children[len - 1]);
      linkToNxt(items[itemsLen - 1], listContainer.children[0]);

      linkToNxt(listContainer.children[len - 1], items[0]);
      linkToPrv(listContainer.children[0], items[itemsLen - 1]);
    } else {
      linkToPrv(items[0], items[itemsLen - 1]);
      linkToNxt(items[itemsLen - 1], items[0]);
    }

    if (itemsLen > 1) {
      linkToNxt(items[0], items[1]);
      linkToPrv(items[itemsLen - 1], items[itemsLen - 2]);

      for (let i = 1; i < itemsLen - 2; i++) {
        linkToPrv(items[i], items[i - 1]);
        linkToNxt(items[i], items[i + 1]);
      }
    }
  };

  NavigationManager.prototype.setFocus = function setFocus(item) {
    item.classList.add('focus');
    item.focus();
    _focusedEl = item;
    this.dispatchEvent('focusChanged', {
      focusedElement: item
    });
  };

  NavigationManager.prototype.getElementByNavId = function getElementByNavId(
    navId
  ) {
    return document.querySelector(`[data-nav-id="${navId}"]`);
  };

  NavigationManager.prototype.switchContext = function switchContext(
    selector,
    id,
    direction
  ) {
    if (!selector) {
      throw Error('selector is undefined');
    }

    const selected = id ? id : this.context[selector];
    this.reset(selector, selected, direction);
  };

  NavigationManager.prototype.debug = function debug(message) {
    if (this.DEBUG) {
      console.log(`${this.CLASS_NAME}: ${message}`);
    }
  };

  NavigationManager.prototype.getFocusedEl = function getFocusedEl() {
    return _focusedEl;
  };

  exports.NavigationManager = new NavigationManager();
})(window);
