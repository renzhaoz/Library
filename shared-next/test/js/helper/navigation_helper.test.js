/* eslint-disable no-undef, global-require, new-cap */
describe('helper js <navigation_helper> test', () => {
  const { wrap } = require('jest-snapshot-serializer-raw');

  beforeAll(done => {
    require('../../../js/helper/navigation/navigation_helper');
    done();
  });

  test('NavigationHelper, HorizontalNavigator, VerticalNavigator, BoxNavigator, GroupNavigator should be object', done => {
    expect(typeof window.NavigationHelper).toBe('object');
    expect(typeof window.HorizontalNavigator).toBe('object');
    expect(typeof window.NavigationHelper).toBe('object');
    expect(typeof window.BoxNavigator).toBe('object');
    expect(typeof window.GroupNavigator).toBe('object');
    done();
  });

  // Test NavigationHelper.resetMenu
  test('NavigationHelper.resetMenu', done => {
    expect(typeof window.NavigationHelper.resetMenu).toBe('function');
    done();
  });

  // Test NavigationHelper.resetMenu return value 1
  test('NavigationHelper.resetMenu should return', done => {
    const spy = jest.spyOn(window.NavigationHelper, 'reset');
    document.body.innerHTML = `
    <form data-subtype="menu">
      <button id="button" />
    </form>
  `;
    const res = window.NavigationHelper.resetMenu();
    expect(spy).toBeCalledTimes(1);
    expect(res).toBeNull();
    done();
  });

  describe('NavigationHelper.resetMenu return value test', () => {
    let spyReset = null;
    let spyUpdateNav = null;
    let spySetFocus = null;
    beforeEach(done => {
      spyReset = jest.spyOn(window.NavigationHelper, 'reset');
      spyUpdateNav = jest.spyOn(window.NavigationHelper, 'updateNav');
      spySetFocus = jest.spyOn(window.NavigationHelper, 'setFocus');
      done();
    });

    // Test NavigationHelper.resetMenu return value 2
    test('NavigationHelper.resetMenu should return focusElement when child > 1', done => {
      document.body.innerHTML = `
        <form data-subtype="menu" id="mainmenu">
          <button id="submenu_button1" class="menu-button"/>
          <button id="submenu_button2" class="menu-button"/>
          <button id="submenu_button3" class="menu-button"/>
        </form>
      `;
      const ele = window.NavigationHelper.resetMenu();
      expect(spyReset).toBeCalledTimes(3);
      expect(spyUpdateNav).toBeCalledTimes(1);
      expect(spySetFocus).toBeCalledTimes(1);
      expect(wrap(ele)).toMatchSnapshot();
      done();
    });

    // Test NavigationHelper.resetMenu return value 3
    test('NavigationHelper.resetMenu should return focusElement when child = 1', done => {
      document.body.innerHTML = `
        <form data-subtype="menu" id="mainmenu">
          <button id="button" class="menu-button"/>
        </form>
      `;
      const ele = window.NavigationHelper.resetMenu();
      expect(spyReset).toBeCalledTimes(1);
      expect(spyUpdateNav).toBeCalledTimes(1);
      expect(spySetFocus).toBeCalledTimes(1);
      expect(wrap(ele)).toMatchSnapshot();
      done();
    });
  });

  // Test NavigationHelper.scrollToElement
  test('NavigationHelper.scrollToElement', done => {
    expect(typeof window.NavigationHelper.scrollToElement).toBe('function');
    done();
  });

  // Test NavigationHelper.scrollToElement return value
  test('NavigationHelper.scrollToElement return value test', done => {
    const element = document.createElement('div');
    const spyWidth = jest
      .spyOn(element, 'offsetWidth', 'get')
      .mockReturnValue(10);
    const spyHeight = jest
      .spyOn(element, 'offsetHeight', 'get')
      .mockReturnValue(10);
    const spyRects = jest
      .spyOn(element, 'getClientRects')
      .mockReturnValue([{ top: 10, bottom: 10 }]);
    window.HTMLElement.prototype.scrollIntoView = jest.fn();
    const event = { key: 'ArrowUp' };
    const top = 20;
    const bottom = 0;
    window.NavigationHelper.scrollToElement(element, event, top, bottom);
    expect(spyWidth).toBeCalledTimes(1);
    expect(spyHeight).toBeCalledTimes(1);
    expect(spyRects).toBeCalledTimes(1);
    expect(window.HTMLElement.prototype.scrollIntoView).toBeCalledTimes(1);
    window.HTMLElement.prototype.scrollIntoView = undefined;
    done();
  });

  // Test NavigationHelper.Key2Dir
  test('NavigationHelper.Key2Dir', done => {
    expect(typeof window.NavigationHelper.Key2Dir).toBe('function');
    done();
  });

  // Test NavigationHelper.Key2Dir return value
  test('NavigationHelper.Key2Dir return value test', done => {
    const left = window.NavigationHelper.Key2Dir('ArrowLeft');
    expect(left).toBe(0);
    const right = window.NavigationHelper.Key2Dir('ArrowRight');
    expect(right).toBe(1);
    const up = window.NavigationHelper.Key2Dir('ArrowUp');
    expect(up).toBe(2);
    const down = window.NavigationHelper.Key2Dir('ArrowDown');
    expect(down).toBe(3);
    const test = window.NavigationHelper.Key2Dir('test');
    expect(test).toBeUndefined();
    done();
  });

  // Test HorizontalNavigator.left, right, up, down
  test('HorizontalNavigator.left, right, up, down should be function', done => {
    expect(typeof window.HorizontalNavigator.left).toBe('function');
    expect(typeof window.HorizontalNavigator.right).toBe('function');
    expect(typeof window.HorizontalNavigator.up).toBe('function');
    expect(typeof window.HorizontalNavigator.down).toBe('function');
    done();
  });

  // Test HorizontalNavigator.left, right, up, down return value
  test('HorizontalNavigator.left, right, up, down return value test', done => {
    const left = window.HorizontalNavigator.left(0);
    expect(left).toBe(-1);
    const right = window.HorizontalNavigator.right(0);
    expect(right).toBe(1);
    const left1 = window.HorizontalNavigator.left(1);
    expect(left1).toBe(0);
    const right1 = window.HorizontalNavigator.right(-1);
    expect(right1).toBe(0);
    const up = window.HorizontalNavigator.up(0);
    expect(up).toBe(0);
    const down = window.HorizontalNavigator.down(0);
    expect(down).toBe(0);
    done();
  });

  // Test BoxNavigator.left, right, up, down
  test('BoxNavigator.left, right, up, down should be function', done => {
    expect(typeof window.BoxNavigator.left).toBe('function');
    expect(typeof window.BoxNavigator.right).toBe('function');
    expect(typeof window.BoxNavigator.up).toBe('function');
    expect(typeof window.BoxNavigator.down).toBe('function');
    done();
  });

  // Test BoxNavigator.left, right, up, down return value
  test('BoxNavigator.left, right, up, down return value test', done => {
    const left = window.BoxNavigator.left(0);
    expect(left).toBe(-1);
    const right = window.BoxNavigator.right(0);
    expect(right).toBe(0);
    const up = window.BoxNavigator.up(0);
    expect(up).toBe(-2);
    const down = window.BoxNavigator.down(0);
    expect(down).toBe(0);
    done();
  });

  // Test GroupNavigator.left, right, up, down
  test('GroupNavigator.left, right, up, down should be function', done => {
    expect(typeof window.GroupNavigator.left).toBe('function');
    expect(typeof window.GroupNavigator.right).toBe('function');
    expect(typeof window.GroupNavigator.up).toBe('function');
    expect(typeof window.GroupNavigator.down).toBe('function');
    done();
  });

  // Test GroupNavigator.left, right, up, down return value
  test('GroupNavigator.left, right, up, down return value test', done => {
    // eslint-disable-next-line no-unused-vars
    GroupNavigator.getGroupInfo = function getGroupInfo(navIndex, isup) {
      return {
        curGroupIndex: 0,
        curGroupLen: 6,
        destGroupLen: 8
      };
    };
    const left = window.GroupNavigator.left(0);
    expect(left).toBe(-1);
    const right = window.GroupNavigator.right(0);
    expect(right).toBe(0);
    const up = window.GroupNavigator.up(0);
    expect(up).toBe(-2);
    const down = window.GroupNavigator.down(0);
    expect(down).toBe(3);
    done();
  });

  afterEach(done => {
    jest.restoreAllMocks();
    done();
  });

  afterAll(done => {
    document.body.innerHTML = '';
    GroupNavigator.getGroupInfo = null;
    done();
  });
});
