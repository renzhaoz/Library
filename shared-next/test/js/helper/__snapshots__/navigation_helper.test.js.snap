// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper js <navigation_helper> test NavigationHelper.resetMenu return value test NavigationHelper.resetMenu should return focusElement when child = 1 1`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="menu-button focus"
    data-nav-id="mainmenu0"
    id="button"
    style="--nav-left: mainmenu0; --nav-right: mainmenu0; --nav-up: mainmenu0; --nav-down: mainmenu0;"
    tabindex="1"
  >
    
        
      
  </button>,
}
`;

exports[`helper js <navigation_helper> test NavigationHelper.resetMenu return value test NavigationHelper.resetMenu should return focusElement when child > 1 1`] = `
Object {
  Symbol(jest-snapshot-serializer-raw): <button
    class="menu-button focus"
    data-nav-id="mainmenu0"
    id="submenu_button1"
    style="--nav-left: mainmenu0; --nav-right: mainmenu0; --nav-up: mainmenu2; --nav-down: mainmenu1;"
    tabindex="1"
  >
    
          
  </button>,
}
`;
